File: Assets/Scripts/MonsterController.js
#pragma strict

public var stunnedMaterial : Material;
private var originalMaterial : Material;

private var target : GameObject;

private var levelGenerator : LevelGenerator;

private var stunnedSince : float;
private var isBeaten : boolean = false;
private var stunTime : float;

function Start () {
	levelGenerator = GameObject.Find("LevelGenerator").GetComponent(LevelGenerator);
	originalMaterial = transform.renderer.material;
	stunTime = levelGenerator.levelParameters().monsterStunTime;
	target = GameObject.Find("Player(Clone)");
}

function Update () {
	if (!target) {
		target = GameObject.Find("Player(Clone)");
	}
	
	if (target) {	
		lookAtTarget();
		if (shouldChase()) moveTowardsTarget();
	}
	
	checkStunnedTime();	
}

function shouldChase () {
	return !stunnedSince && !isBeaten;
}

function stun () {
	stunnedSince = Time.time;
	transform.renderer.material = stunnedMaterial;
}

function unStun () {
	stunnedSince = 0;
	transform.renderer.material = originalMaterial;
}

function checkStunnedTime() {
	if (Time.time - stunnedSince >= stunTime)
		unStun();
}

function setBeaten (beaten : boolean) {
	isBeaten = beaten;
}

function lookAtTarget () {
	transform.LookAt(target.transform);
}

function moveTowardsTarget () {
	var controller : CharacterController = GetComponent(CharacterController);
	
	var direction : Vector3 = (target.transform.localPosition - transform.localPosition).normalized;

	controller.Move(
		direction * levelGenerator.levelParameters().monsterSpeed * Time.deltaTime
	);
	transform.localPosition.y = transform.localScale.y / 2;
}

File: Assets/Scripts/MonsterController.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class MonsterController : MonoBehaviour
{

    public Material stunnedMaterial;
    private Material originalMaterial;

    private GameObject target;

    private LevelGenerator levelGenerator;

    private float stunnedSince;
    private bool isStunned;
    private bool isBeaten = false;
    private float stunTime;

    void Start()
        {
            levelGenerator = GameObject.Find("levelGenerator").GetComponent<LevelGenerator>();
            originalMaterial = transform.GetComponent<Renderer>().material;
            stunTime = levelGenerator.LevelParameters().monsterStunTime;
            target = GameObject.Find("Player(Clone)");
        }

    void Update()
        {
            if (!target)
            {
                target = GameObject.Find("Player(Clone)");
            }

            if (target)
            {
                LookAtTarget();
                if (ShouldChase()) MoveTowardsTarget();
            }

            CheckStunnedTime();
        }

        public bool ShouldChase()
        {
            return !isStunned && !isBeaten;
        }

        public void Stun()
        {
            isStunned = true;
            stunnedSince = Time.time;
            transform.GetComponent<Renderer>().material = stunnedMaterial;
        }

        public void UnStun()
        {
            stunnedSince = 0;
            isStunned = false;
            transform.GetComponent<Renderer>().material = originalMaterial;
        }

        public void CheckStunnedTime()
        {
            if (Time.time - stunnedSince >= stunTime)
                UnStun();
        }

        public void SetBeaten(bool beaten)
        {
            isBeaten = beaten;
        }

        public void LookAtTarget()
        {
            transform.LookAt(target.transform);
        }

        public void MoveTowardsTarget()
        {
            Vector3 direction = (target.transform.localPosition - transform.localPosition).normalized;
            GetComponent<CharacterController>().Move(
                direction * levelGenerator.LevelParameters().monsterSpeed * Time.deltaTime
            );
            transform.localPosition.Set(transform.localScale.x, transform.localScale.y / 2, transform.localScale.z);
        }

}
